Implement UTXOSetValue for freezing

MichaelKim20 opened the pull request:
Add freezing capability to UTXO #204 of part

UTXOSetValue is used by UTXOSet
To freeze UTXO, add the following information to UTXOSet:
The height of block.
The type of transaction.

![UTXO 008](https://user-images.githubusercontent.com/11639939/65367117-899caf00-dc67-11e9-87fb-90c0f0c91b15.jpeg)

![Block_Tx_UTXO](https://user-images.githubusercontent.com/11639939/65367140-e7c99200-dc67-11e9-9187-796bec123aa9.gif)

<img width="920" alt="Transaction & UTXO in Freeze" src="https://user-images.githubusercontent.com/11639939/65112733-9c22a880-da1b-11e9-9f1b-55f4b00098c5.png">

AndrejMitrovic commented:
This pull request introduces `findUTXOSetValue`, and then removes it in another commit. It should just be removed from the commit it was originally added in, and then the commit `Remove unused function UTXOSetValue.findUTXOSetValue` can be removed.

Geod24 commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -27,6 +27,51 @@ import std.file;
+public struct UTXOSetValue
+{
+    /// Height of the block to be activated
+    ulong height;
```
This could use a more descriptive name.
I can suggest "height_lock", "frozen_until", "locked_until", "min_height"... But I'm open to suggestion @TrustHenry @AndrejMitrovic @MukeunKim

AndrejMitrovic replied:
I suggest snake_case.

TrustHenry replied:
I suggest "height_lock".

MichaelKim20 replied:
I suggest "locked_until"~

AndrejMitrovic replied:
I would go with @MukeunKim's suggestion, or maybe `unlock_height`

MichaelKim20 replied:
I like `unlock_height`

TrustHenry replied:
Me too `unlonk_height`  because  `height + 1;`

Geod24 commented:
> This pull request introduces findUTXOSetValue, and then removes it in another commit.b

Damn Github fix your commit ordering.

AndrejMitrovic commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
+                this.utxo_db.remove(utxo_hash);
+            }
+
+            ulong actived_height = height + 1;
```
Why +1? The UTXO will be in the block at the provided height that's in the parameter, not height + 1.

MichaelKim20 replied:
Next Block height

MichaelKim20 replied:
I entered the available height, i.e. unlock height.
I wanted to specify what could be used in the next block, not in the current block.

AndrejMitrovic commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
+            foreach (const ref input; tx.inputs)
+            {
+                if (!this.getUTXOSetValue(input.previous, input.index, utxo_value))
+                    return;
```
These should `assert(0)`, not fail silently.

MukeunKim replied:
I use log.error()

MichaelKim20 replied:
Remove this part. I have confirmed that this is not necessary.

AndrejMitrovic commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
+                if (!this.getUTXOSetValue(input.previous, input.index, utxo_value))
+                    return;
+                else if (utxo_value.type != TxType.Payment)
+                    return;
```
Same comment about assert.

MukeunKim replied:
I use log.error()

AndrejMitrovic commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -81,25 +126,116 @@ public class UTXOSet
+                else if (n_freeze == tx.inputs.length)
+                    tx_type = TxType.Freeze;
+                else
+                    return;
```
`assert(0)`

MichaelKim20 commented:
Ready fo Review

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -542,7 +543,138 @@ unittest
+                    assert(ledger.acceptTransaction(tx) == is_valid);
+                });
+
+            if (is_valid)
```
`is_valid` is always true, because it is never set to false.

MichaelKim20 replied:
It's not used here, so we'll have to remove it

MichaelKim20 replied:
However, if an error occurs and the block is not created, then the values of these variables must not be changed.

MichaelKim20 replied:
Ready for review

AndrejMitrovic commented:
Unless I'm missing something, this is what `updateUTXOCache` should look like:

```D
public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
{
    ulong unlock_height = (tx.type == TxType.Freeze)
        ? height + 2016 : height + 1;

    foreach (const ref input; tx.inputs)
    {
        auto utxo_hash = getHash(input.previous, input.index);
        this.utxo_db.remove(utxo_hash);
    }

    Hash tx_hash = tx.hashFull();
    foreach (idx, output; tx.outputs)
    {
        auto utxo_hash = getHash(tx_hash, idx);
        const UTXOSetValue value = UTXOSetValue(unlock_height, tx.type, output);
        this.utxo_db[utxo_hash] = value;
    }
}
```

And also, this code passes the test-suite you wrote.

The validation rules for payment => freeze => melting should live in `isInvalidReason`, and not be scattered across multiple functions. `updateUTXOCache()`'s only purpose is to Remove spent UTXOs, and add new ones.

AndrejMitrovic commented:
My previous comment was wrong with regards to unfreezing. Let me see if it can still be simplified..

AndrejMitrovic commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -81,25 +126,105 @@ public class UTXOSet
+                        n_payment++;
+                }
+
+                if (n_freeze == tx.inputs.length)
```
This is the part I find a bit confusing. The new outputs (UTXOs) will be be melting only if **all** of the inputs are frozen UTXOs. But if one of the was a Payment, then all of the outputs will not be melting, they will be instantly melted?

MichaelKim20 commented:
### There are three more commit after that.
It is PR #326.

**Includes:**
Implement validation of transactions associated with freezing
Add test for freezing and melting while creating blocks

MichaelKim20 commented:
Ready for review

AndrejMitrovic commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -81,25 +126,89 @@ public class UTXOSet
+
+            foreach (const ref input; tx.inputs)
+            {
+                auto utxo_hash = getHash(input.previous, input.index);
```
This part seems to be the same for the freeze transactions too. Maybe it could be moved before the `if (tx.type == TxType.Freeze)` check so the code doesn't have to be duplicated.

AndrejMitrovic commented:
I suggest rewriting these two functions as follows:

```D

/***************************************************************************

    Add all of a transaction's outputs to the Utxo set,
    and remove the spent outputs in the transaction from the set.

    Params:
        tx = the transaction
        height = Height of the block where UTXO will be unlocked

***************************************************************************/

public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
{
    import std.algorithm : any;

    foreach (const ref input; tx.inputs)
    {
        auto utxo_hash = getHash(input.previous, input.index);
        this.utxo_db.remove(utxo_hash);
    }

    // defaults to next block
    ulong unlock_height = height + 1;

    // for payments of frozen transactions, it will melt after 2016 blocks
    if (tx.type == TxType.Payment
        && tx.inputs.any!(input => this.getUTXOSetValue(
            input.previous, input.index).type == TxType.Freeze))
    {
        unlock_height = height + 2016;
    }

    Hash tx_hash = tx.hashFull();
    foreach (idx, output; tx.outputs)
    {
        auto utxo_hash = getHash(tx_hash, idx);
        auto utxo_value = UTXOSetValue(unlock_height, tx.type, output);
        this.utxo_db[utxo_hash] = utxo_value;
    }
}

/***************************************************************************

    get an UTXOSetValue in the UTXO set.

    Params:
        hash = the hash of transation
        index = the index of the output
        value = will contain the UTXOSetValue if found

    Return:
        Return true if the UTXO was found

***************************************************************************/

private UTXOSetValue getUTXOSetValue (Hash hash, size_t index)
    nothrow @safe
{
    auto utxo_hash = getHash(hash, index);

    UTXOSetValue value;
    if (this.utxo_db.find(utxo_hash, value))
        assert(0);  // the UTXO must exist
    return value;
}
```

It gets rid of code duplication. And also, `getUTXOSetValue` must return a valid value because `updateUTXOCache` is called after validation. We know at this point a previous UTXO must exist.

MichaelKim20 commented:
To use 'any!' I think it's better to change it.

MichaelKim20 committed:
Modify delegate UTXOFinder for freezing

Use `UTXOSetValue` as a parameter instead of `Output`.

Co-Authored-By: HyeonyeobKim <hyeonyeob.kim@bpfkorea.org>

MichaelKim20 committed:
Implement UTXOSetValue for freezing

UTXOSetValue is used by UTXOSet
To freeze UTXO, add the following information to UTXOSet:
The height of block.
The type of transaction.

Co-Authored-By: HyeonyeobKim <hyeonyeob.kim@bpfkorea.org>

MichaelKim20 committed:
Apply UTXOSetValue to UTXOSet

Co-Authored-By: HyeonyeobKim <hyeonyeob.kim@bpfkorea.org>

MichaelKim20 commented:
Ready for review

AndrejMitrovic commented:
The bug in my suggestion was the over-eager removal of inputs. So this can be fixed to this:

```D
public void updateUTXOCache (const ref Transaction tx, ulong height) @safe
{
    import std.algorithm : any;

    // defaults to next block
    ulong unlock_height = height + 1;

    // for payments of frozen transactions, it will melt after 2016 blocks
    if (tx.type == TxType.Payment
        && tx.inputs.any!(input => this.getUTXOSetValue(
            input.previous, input.index).type == TxType.Freeze))
    {
        unlock_height = height + 2016;
    }

    Hash tx_hash = tx.hashFull();
    foreach (idx, output; tx.outputs)
    {
        auto utxo_hash = getHash(tx_hash, idx);
        auto utxo_value = UTXOSetValue(unlock_height, tx.type, output);
        this.utxo_db[utxo_hash] = utxo_value;
    }

    foreach (const ref input; tx.inputs)
    {
        auto utxo_hash = getHash(input.previous, input.index);
        this.utxo_db.remove(utxo_hash);
    }
}
```

AndrejMitrovic commented:
However, this assertion will fail:

```D
private UTXOSetValue getUTXOSetValue (Hash hash, size_t index)
    nothrow @safe
{
    auto utxo_hash = getHash(hash, index);

    UTXOSetValue value;
    if (this.utxo_db.find(utxo_hash, value))
        assert(0);  // assertion triggers
    return value;
}
```

That means either something with the tests is wrong, or the validation functions aren't catching an edge-case.

MichaelKim20 commented:
@AndrejMitrovic I fixed it. Thank you.

MichaelKim20 commented:
The full test code using `UTXSet.updateUTXOCache` is in PR #326.
So we had a hard time proceeding with the review here.
Anyway, thank you all.
